### DISCUSSION TITLE
Run locally on desktop rather than online colab?

### DISCUSSION THREAD

**Discussion by cragginstylie** [nateraw/stable-diffusion-videos#137] [Q&A]
I've tried running this locally on a Windows 11 desktop with Nvidia 3060/12GB. It installs ok, but when I try the "Run locally" python script described in the README, I get the following error:

`AssertionError: Torch not compiled with CUDA enabled`

I'm running Auto1111 with no issues, and I'm confident that torch must be compiled with CUDA support there as well. Maybe the requirements aren't defined correctly to insure the correct binary for Torch is installed? I did the 'pip install -U stable_diffusion_videos` as directed...

I'm also hoping that this repo could be installed, similar to Auto1111, with a 'git clone' or 'git pull', followed by having a launch script (.bat and/or .sh) rather than going the package route.

Any tips would be appreciated. Also willing to help to make my suggestions possible, just need some guidance.

**Comment by nateraw** [comment]
hmm. I'm not sure (I don't have a windows machine, and even if I set up a windows machine in CI here, it won't have a GPU)

I think as long as you have SOME environment with torch correctly installed and working with GPU, it should be fine (i.e. you get `True` when you run `import torch` followed by `torch.cuda.is_available()` within your env). 

I haven't worked with GPU in a long time (maybe 3 years) so I am sorry I'm unable to help more 😢 . Please let me know if you resolve this, and any steps you had to take. perhaps we can use that to update some instructions for windows users. I would appreciate that a lot.

**Reply by cragginstylie** [reply]
Ok, thanks for the reply. The is_available() method returns False, which frankly I expected. Given this works for Auto1111, I think there's 2 instances of Torch installed, one with CUDA support compiled in, and one that doesn't. The one that doesn't was most likely installed along with your package. The next step for me is to try to find the Torch package instance (it's path within the python venv) that was installed with Auto1111. Once I've done that, I may be able to figure out how to use that instance for your package.

Ultimately, I think it comes down to the requirements.txt file within your package isn't pulling a version of Torch compiled with CUDA support. It may take me a bit to figure this out, but if/when I do - I'll reply back here with my findings.

Thanks!

**Reply by nateraw** [reply]
are you using an environment manager or anything? (conda, venv, etc)

I personally use miniconda, which makes it easier to navigate these issues imo. You are right, if you just do simple `pip install stable-diffusion-videos` w/o torch already properly installed, you likely will get a bad install.

My suggestion would be to install torch correctly in some new black environment using the directions in the [getting started page on the pytorch site](https://pytorch.org/get-started/locally/). then, once you can successfully run the `torch.cuda.is_available()` fn and get `True` back, install the package here in the same environment. 

I actually don't have torch in the deps directly here, as I was leaving it up to users to install it correctly (this is _extremely_ common practice for repos/python deps that use torch). It only installs because it's [listed upstream ](https://github.com/huggingface/diffusers/blob/9b37ed33b5fa09e594b38e4e6f7477beff3bd66a/setup.py#L108)in `diffusers`.

As long as you have torch>=1.4 installed and working properly, the install here then (really the diffusers one), will have no effect and should work fine.

Phew...sorry for wall of text, just tryin to give you context haha.

**Reply by nateraw** [reply]
I suggest probably torch 1.13.1, the latest stable version btw.

I'm very sorry, I know how difficult it is to deal with these cuda related issues, especially locally and not on some nice environment already set up for you on a virtual machine like on GCP/paperspace/Azure

**Reply by cragginstylie** [reply]
No worries on "the wall of text". Better to over-communicate anyway. ;-)

Yes, Auto1111 sets up a python venv and installs package deps there. If I run these commands:

`f:`
`cd stable-diffusion-webui\venv`
`set VENV_DIR=.`
`Scripts/Python.exe`
`>>> import torch`
`>>> torch.cuda.is_available()`
`True`
`>>>`

This shows torch with CUDA support is in fact installed, but only within this python venv space. Executing python from anywhere else, the torch.is_available() method returns false.

Here is the torch installer command that Auto1111 uses once the venv has been set:

`pip install torch==1.12.1+cu113 torchvision==0.13.1+cu113 --extra-index-url https://download.pytorch.org/whl/cu113`

Letting upstream diffusers install torch is probably not the best way, IMHO, because it obviously doesn't work. At least not on Windows, and like you - but opposite - I don't have a Linux box with compatible GPU. (Thinking about acquiring one, but I digress.)

I'm wondering out loud... if we were to add a line to the reqs.txt just above the diffusers entry with another entry for torch with the correct version & extensions (i.e. torch=1.12.1+cu113) then it'll probably fix it. Even better would be to use an installer/launch script similar to Auto1111 to set up a python venv with which to isolate all the package deps for better compartmentalization.

I'll have to uninstall this package, and then try a git clone on the repo so I can test this idea.

**Reply by cragginstylie** [reply]
Hey, brother - no need to apologize at all! THANK YOU for even providing this package! Happy to contribute back in my own small way!

**Reply by nateraw** [reply]
haha all good :) happy to try and help. tldr it sounds like if you just run the python from that venv you created with automatic, and install `stable-diffusion-videos` within it, you should be good to go. 

---

Now, with regards to your suggestions:

> Letting upstream diffusers install torch is probably not the best way, IMHO, because it obviously doesn't work

I think you'll find that this is the standard and accepted approach when it comes to torch. No need to add complexity (I'm referring to bat files in the automatic repo) to the project here, when the real issue is upstream at torch.

> I'm wondering out loud... if we were to add a line to the reqs.txt just above the diffusers entry with another entry for torch with the correct version & extensions (i.e. torch=1.12.1+cu113) then it'll probably fix it.

I realize you're spitballin, which I appreciate, but this approach would cause more issues than it would solve:

- It doesn't take into consideration folks might not be using it on CUDA devices (this repo supports MPS on mac for example, and hopefully TPU at some point).
- That would be a hard coded version of cuda, which we definitely can't assume
- It's a very strict torch dep, whereas the current is anything `>1.4.0`

The repo you are referring to ends up running `launch.py` which is basically just installing stuff in a venv. You can easily do this yourself 😄 (if you don't know how, I PROMISE its worth taking a couple hours to dive deep into venv).

**Reply by nateraw** [reply]
> Even better would be to use an installer/launch script similar to Auto1111 to set up a python venv with which to isolate all the package deps for better compartmentalization.

I'm incredibly uncomfortable with doing this, and really would prefer not to, but could be convinced if:

a. Someone with a windows machine does it (not me)
b. it doesn't add too much unnecessary complexity to the repo

**Reply by cragginstylie** [reply]
> The repo you are referring to ends up running launch.py which is basically just installing stuff in a venv. You can easily do this yourself 😄 (if you don't know how, I PROMISE its worth taking a couple hours to dive deep into venv).

I know python well, and use venv frequently; however, mostly on linux - not windows. I'm using this windows-based gaming tower as my past experience was always that the newest video card drivers were often unavailable or unstable on linux, especially 3D drivers. With linux so much more mainstream now, I probably could've wiped windows 11, installed the latest Ubuntu and had no problems. I digress...

> I'm incredibly uncomfortable with doing this, and really would prefer not to, but could be convinced if:

Well.....
a. I have a windows machine and *can* do it
b. which would be your call as to whether it adds too much complexity or not

Using python's venv is always a best practice, IMO. I assume this is a big reason you stated:

> I PROMISE its worth taking a couple hours to dive deep into venv

LOL

Seriously, tho... I'll take a stab at it.

**Reply by nateraw** [reply]
haha really give the venv thing a try and let me know how it goes...then we can go from there.

I seriously bet you don't even need a new one if you already have the existing automatic111 one. Probably as simple as a call to the `venv/.../Python.exe -m pip install stable-diffusion-videos` then run whatever script you write with the library using the same executable... `venv/.../Python.exe your_script.py`. (the ... here means I don't know what the exact path would be).

**Reply by cragginstylie** [reply]
Exactly my plan. Thanks again for everything, especially your time!

**Comment by Raketemensch** [comment]
I'm not sure if you got this working, but I was able to get pytorch installed using Cuda 11.7 -- I don't think there's a version of pytorch that uses cuda 12 yet.

After pip installing stable-diffusion-videos (and the separate cuda 11.7 install) I then had to:

```
pip uninstall torch torchvision torchaudio
pip install torch torchvision torchaudio --extra-index-url https://download.pytorch.org/whl/cu117

```

This will install versions of the torch wheels built for cuda 11.7. Everything is working locally in Windows 10 for me now. I'd rather run it in Linux, but the good video card will sadly always be in a Windows machine...

**Comment by UnconnectedBedna** [comment]
Sorry for bump of necro thread.

What version of python is recommended for the pyenv?
I do not want to use the same as my stable diffusion but rather an isolated pyenv version.
(Idc about bloating my system)

My SD is running under 3.10.6, not the newest but it is what they recommend.
